Migrate status

GaryQian opened the pull request:
`flutter migrate status` subcommand.

This command shows the diffs staged in the migrate working directory and the manifest of files to be changed.

Depends on https://github.com/flutter/flutter/pull/101937

GaryQian committed:
Migrate status

GaryQian committed:
Add tests

GaryQian committed:
cleanup logging

blasten commented on packages/flutter_tools/lib/src/commands/migrate_status.dart:
```diff
@@ -0,0 +1,163 @@
+    Directory workingDirectory = project.directory.childDirectory(kDefaultMigrateWorkingDirectoryName);
+    final String? customWorkingDirectoryPath = stringArg('working-directory');
+    if (customWorkingDirectoryPath != null) {
+      if (customWorkingDirectoryPath.startsWith(fileSystem.path.separator) || customWorkingDirectoryPath.startsWith('/')) {
```
how does this work on Windows?

blasten commented on packages/flutter_tools/lib/src/commands/migrate_status.dart:
```diff
@@ -0,0 +1,163 @@
+      'working-directory',
+      help: 'Specifies the custom migration working directory used to stage and edit proposed changes. '
+            'This path can be absolute or relative to the flutter project root.',
+      valueHelp: 'path',
```
should it mention the default directory?

GaryQian replied:
Yes, ill add it in the help text!

christopherfujino replied:
Idk if this will be painful, but can we name this something else, maybe `workspace-directory`? I think this will be confusing with the current working directory of the tool invocation. In reading the code, I thought this field was a way to override the project directory.

GaryQian replied:
Renamed to `migrate_staging_dir`

blasten commented on packages/flutter_tools/lib/src/commands/migrate_status.dart:
```diff
@@ -0,0 +1,163 @@
+      help: 'Shows the diff output when enabled. Enabled by default.',
+    );
+    argParser.addFlag(
+      'show-added-files',
```
what is the use case for this flag?

GaryQian replied:
Many of these flags assist in spam-reduction to make it easier to find the diffs/info the user is actually looking for.

blasten commented on packages/flutter_tools/lib/src/commands/migrate_status.dart:
```diff
@@ -0,0 +1,163 @@
+      valueHelp: 'path',
+    );
+    argParser.addFlag(
+      'diff',
```
what is the use case for disabling this flag?

GaryQian replied:
The diffs can often be quite spammy, programmatic uses especially may not wish to have such an extensive output.

blasten commented on packages/flutter_tools/lib/src/commands/migrate_status.dart:
```diff
@@ -0,0 +1,163 @@
+    );
+    argParser.addOption(
+      'project-directory',
+      help: 'The root directory of the flutter project.',
```
for simplicity, should it only assume the current directory is the project directory?

GaryQian replied:
This tool is also intended to be used programmatically, where this option helps immensely.

christopherfujino replied:
It looks like if this is not provided we default to the current working directory, can we document this?

blasten commented on packages/flutter_tools/lib/src/commands/migrate_status.dart:
```diff
@@ -0,0 +1,163 @@
+  Future<Set<DevelopmentArtifact>> get requiredArtifacts async => const <DevelopmentArtifact>{};
+
+  /// Manually marks the lines in a diff that should be printed unformatted for visbility.
+  final Set<int> _initialDiffLines = <int>{0, 1};
```
what do these lines contain?

GaryQian replied:
These lines just include the file names being diffed for example: 
```
diff --git a/Users/garyq/projects/vanillabase1226/android/app/src/main/res/values/styles.xml b/Users/garyq/projects/vanillabase1226/migrate_working_dir/android/app/src/main/res/values/styles.xml
index 1f83a33..8dbda4a 100644
```

christopherfujino replied:
Can you flesh out the dartdoc with this example?

blasten commented on packages/flutter_tools/lib/src/commands/migrate_status.dart:
```diff
@@ -0,0 +1,163 @@
+          int lineNumber = -1;
+          for (final String line in result.diff!.split('\n')) {
+            lineNumber++;
+            if (line.startsWith('---') || line.startsWith('+++') || line.startsWith('&&') || _initialDiffLines.contains(lineNumber)) {
```
what happens if a line starts with `---`, `+++`, `&&` in the original file?

GaryQian replied:
The first character of the line is reserved for the + and - character, so lines that start with those strings in the original files would appear as ` ---` ` +++` and ` &&`. The initial space character would cause the startsWith check to fail.

GaryQian committed:
Merge with utils and manifest

GaryQian committed:
Fix analyze

GaryQian committed:
windows drive regex

GaryQian committed:
Fix regex

GaryQian committed:
Fix null safety

GaryQian committed:
fix typo

christopherfujino commented on packages/flutter_tools/lib/src/commands/migrate.dart:
```diff
@@ -2,20 +2,28 @@
+    bool verbose = false,
     required this.logger,
-    // TODO(garyq): Add each parameters in as subcommands land.
+    // TODO(garyq): Add each of these back in as they land.
```
add in what?

christopherfujino commented on packages/flutter_tools/test/integration.shard/migrate_status_test.dart:
```diff
@@ -0,0 +1,202 @@
+import 'package:flutter_tools/src/base/file_system.dart';
+import 'package:flutter_tools/src/base/logger.dart';
+import 'package:flutter_tools/src/base/platform.dart';
+// import 'package:flutter_tools/src/base/terminal.dart';
```
```suggestion
```

christopherfujino commented on packages/flutter_tools/test/integration.shard/migrate_status_test.dart:
```diff
@@ -0,0 +1,202 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
can you move this file to `packages/flutter_tools/test/command.shard/permeable/migrate_status_test.dart`?

christopherfujino commented on packages/flutter_tools/test/integration.shard/migrate_status_test.dart:
```diff
@@ -0,0 +1,202 @@
+      ]
+    );
+    expect(logger.statusText, contains('''
+Newly addded file at added.file:
```
spelling?

christopherfujino commented on packages/flutter_tools/test/integration.shard/migrate_status_test.dart:
```diff
@@ -0,0 +1,202 @@
+deleted_files:
+''');
+
+    expect(appDir.childFile('lib/main.dart').existsSync(), true);
```
is this needed? we explicitly created it on line 55, are we testing it didn't get deleted?

christopherfujino commented on packages/flutter_tools/test/integration.shard/migrate_status_test.dart:
```diff
@@ -0,0 +1,202 @@
+    conflictFile.createSync(recursive: true);
+    conflictFile.writeAsStringSync('''
+line1
+<<<<<<< /conflcit/conflict.file
```
spelling?

christopherfujino commented on packages/flutter_tools/test/integration.shard/migrate_status_test.dart:
```diff
@@ -0,0 +1,202 @@
+    expect(logger.statusText, contains('''
+@@ -1,3 +1,7 @@
+ line1
++<<<<<<< /conflcit/conflict.file
```
spelling?

GaryQian committed:
Address comments

GaryQian committed:
Rename working dir to staging dir

GaryQian committed:
Cleanup

GaryQian committed:
Analyzer

GaryQian commented:
Ready for another round. Thanks! @christopherfujino

GaryQian committed:
Absolute path

GaryQian commented on packages/flutter_tools/lib/src/commands/migrate_status.dart:
```diff
@@ -0,0 +1,168 @@
+    Directory workingDirectory = project.directory.childDirectory(kDefaultMigrateWorkingDirectoryName);
+    final String? customWorkingDirectoryPath = stringArg('working-directory');
+    if (customWorkingDirectoryPath != null) {
+      if (customWorkingDirectoryPath.startsWith(fileSystem.path.separator) || customWorkingDirectoryPath.startsWith(RegExp(r'[A-Z]:\\'))) {
```
Yep, already did it, just needed to sync it with the other PRs

GaryQian commented:
@zanderso @christopherfujino Anything else that needs to be done here?

zanderso commented:
@GaryQian it looks like there might be one or two unaddressed comments from @christopherfujino

GaryQian commented:
@zanderso  Github UI won't let me comment on those apparently,

I already renamed the working directory to be `migrate_staging_dir` and added docs explaining the first two line handling.

blasten reviewed (approved):
LGTM

GaryQian committed:
Line lengths

GaryQian committed:
Split migrate subcommend into lines

GaryQian committed:
Sync changes

GaryQian committed:
Sync staging rename
